File: marsrover/src/main/grid/GridSurface.cpp
#include "GridComponent.cpp"
#include "ObstacleCell.cpp"

class GridSurface : public GridComponent
{
private:
    int xUpperCoordinate;
    int yUpperCoordinate;
    int xLowerCordinate = 0;
    int yLowerCoordinate = 0;
    set<GridComponent> obstacles;

public:
    GridSurface(int xUpperCoordinate, int yUpperCoordinate)
    {
        this->xUpperCoordinate = xUpperCoordinate;
        this->yUpperCoordinate = yUpperCoordinate;
    }
    void addObstacles(ObstacleCell obstacle)
    {
        obstacles.insert(obstacle);
    }
    void display()
    {
        for (auto it : obstacles)
        {
            it.display();
        }
    }
    set<GridComponent> getObstacles()
    {
        return this->obstacles;
    }
    void setXUpperCoordinate(int xUpperCoordinate)
    {
        this->xUpperCoordinate = xUpperCoordinate;
    }
    void setYUpperCoordinate(int yUpperCoordinate)
    {
        this->yUpperCoordinate = yUpperCoordinate;
    }
    int getXUpperCoordinate()
    {
        return this->xUpperCoordinate;
    }
    int getYUpperCoordinate()
    {
        return this->yUpperCoordinate;
    }
    int getXLowerCoordinate()
    {
        return this->xLowerCordinate;
    }
    int getYLowerCoordinate()
    {
        return this->yLowerCoordinate;
    }
};

File: marsrover/src/main/java/com/thoughtworks/grid/GridSurface.java
package java.com.thoughtworks.grid;

import java.util.*;
// import java.util.Iterator;

public class GridSurface implements GridComponent {

    private final int xUpperCoordinate;
    private final int yUpperCoordinate;
    private final int xLowerCordinate = 0;
    private final int yLowerCoordinate = 0;
    private Set<GridComponent> obstacles;

    public GridSurface(int xUpperCoordinate, int yUpperCoordinate) {
        this.xUpperCoordinate = xUpperCoordinate;
        this.yUpperCoordinate = yUpperCoordinate;
    }

    public void addObstacles(GridComponent obstacle) {
        obstacles.add(obstacle);
    }

    @Override
    public void display() {
        // Iterator iter = obstacles.iterator();
        // while (iter.hasNext()) {
        // iter.next();
        // }
    }

    public Set<GridComponent> getObstacles() {
        return this.obstacles;
    }

    public int getXUpperCoordinate() {
        return this.xUpperCoordinate;
    }

    public int getYUpperCoordinate() {
        return this.yUpperCoordinate;
    }

    public int getXLowerCoordinate() {
        return this.xLowerCordinate;
    }

    public int getYLowerCoordinate() {
        return this.yLowerCoordinate;
    }
};
